### DISCUSSION TITLE
手拿食物时无法使用左键对应功能

### DISCUSSION THREAD

**Discussion by WhiteElephant-abc** [TouchController/TouchController#115] [输入控制逻辑]
在国际基岩版1.21.51.02中，手拿食物仍然可以挖掘方块与攻击实体

**Comment by fifth-light** [comment]
测试在 PE 版 v0.14.3 中挖方块不行，但是可以攻击实体。

使用 TouchController 手拿食物不可以破坏方块，但是可以攻击实体，行为一致。

这又是一个新老版本不一致的问题了。

**Comment by fifth-light** [comment]
或许可以做成一个开关，让可使用物品长按方块时依然破坏方块

**Reply by Mooling0602** [reply]
> 或许可以做成一个开关，让可使用物品长按方块时依然破坏方块

食物方面可以检测饱食度解决，匹配基岩版行为；盾的话可以在长按没有对准方块时触发格挡，对准方块时触发破坏

**Reply by fifth-light** [reply]
逐个适配物品的话有很大问题：原版物品我可以列举所有情况，但是 mod 物品就很难了。因此我更希望有一种统一的方法可以处理所有物品；退一步来说，即使没有一个统一的方法，我也至少需要有一个逻辑让玩家手动配置相应的物品列表，以方便使用其他 mod 的玩家添加 mod 物品到相应列表中。

**Reply by fifth-light** [reply]
其实核心问题就是 Minecraft Java 版在代码中没有一个方法可以在不执行相应操作的前提下检测物品是否可以使用。举例来说就是在代码中你无法在不实际吃食物的前提下判断食物是否可以食用，无法在不实际射箭的前提下判断弓箭是否可以使用。我知道当然可以自己写各种判断代码来实现这些逻辑，但是相对来说比较麻烦，并且难以为各种 mod 的物品实现相应的逻辑（例如需要充电的物品有电才能使用等）。

**Reply by WhiteElephant-abc** [reply]
> 其实核心问题就是 Minecraft Java 版在代码中没有一个方法可以在不执行相应操作的前提下检测物品是否可以使用。举例来说就是在代码中你无法在不实际吃食物的前提下判断食物是否可以食用，无法在不实际射箭的前提下判断弓箭是否可以使用。我知道当然可以自己写各种判断代码来实现这些逻辑，但是相对来说比较麻烦，并且难以为各种 mod 的物品实现相应的逻辑（例如需要充电的物品有电才能使用等）。

但是像苹果皮，是怎么判断的呢，硬编码物品列表吗？显然不是，苹果皮实际上适配mod物品

**Reply by Mooling0602** [reply]
> 逐个适配物品的话有很大问题：原版物品我可以列举所有情况，但是 mod 物品就很难了。因此我更希望有一种统一的方法可以处理所有物品；退一步来说，即使没有一个统一的方法，我也至少需要有一个逻辑让玩家手动配置相应的物品列表，以方便使用其他 mod 的玩家添加 mod 物品到相应列表中。

关于物品列表我有一些想法：
- 主动适配原版物品（内建配置文件）
- 支持添加Mod物品（用户修改配置文件）
- 支持Mod主动往TouchController添加物品（接口）
